atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainer.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.cassandra;
+
+import com.google.errorprone.annotations.CompileTimeConstant;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.refreshable.Disposable;
+import com.palantir.refreshable.Refreshable;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * This container creates new resources when the underlying refreshable changes.
+ * After a refresh, previous resources are closed to avoid resource leaks.
+ *
+ * There is no guarantee that a resource provided via {@link #get} will <i>not</i> be closed whilst in use.
+ * Consumers should be resilient to any uses after the resource has been closed, and where necessary, retry.
+ *
+ * After {@link #close()} has returned, no further resources will be created, and all managed resources will be
+ * closed.
+ */
+public final class ReloadingCloseableContainer<T extends AutoCloseable> implements AutoCloseable {
+    private static final SafeLogger log = SafeLoggerFactory.get(ReloadingCloseableContainer.class);
+
+    private final AtomicReference<Optional<T>> currentResource;
+    private final Refreshable<T> refreshableResource;
+    private final Disposable refreshableSubscriptionDisposable;
+    private final ReadWriteLock isClosedLock;
+
+    @GuardedBy("isClosedLock")
+    private volatile boolean isClosed;
+
+    private <K> ReloadingCloseableContainer(Refreshable<K> refreshableFactoryArgument, Function<K, T> factory) {
```

Comment:
`refreshableFactoryArgument` is a bit of a weird name - this is probably more `refreshableParameter` or something